### PR TITLE
Overledger New actions added (read from smart contract and sign a transaction) and modification of existing Actions (Prepare and execute transactions)  - tied to issue ticket submitted to Pipedream

### PR DESCRIPTION
**Pull request by philbuuza**
The Overledger Pipedream Actions did not have the correct properties in order to work with the the Quant Overledger API request parameters. As well as this other actions added to complete the API sequence. We want the Overledger app in pipedream to be able to perform the the read, monitoring and writing actions.  please have a look at the issue submitted: 

https://github.com/PipedreamHQ/pipedream/issues/14150.

Modifications have been to the prepare and execute actions as these did it seem to be functional.

Phil Buuza

## Release Notes

- **New Features**
  - Introduced new modules for signing transactions and reading from smart contracts.
  - Added a property to specify the environment in transaction preparation and execution.
  - Enhanced functionality for preparing smart contract transactions with improved parameter handling.

- **Improvements**
  - Updated descriptions for transaction properties to enhance clarity.
  - Streamlined transaction preparation and signing processes with consolidated request bodies.
  - Added constants for mapping Overledger environments and technology options to token symbols.
  - Improved flexibility by allowing selection between different Overledger environments.

- **Bug Fixes**
  - Corrected a typo in the summary method for account events.

- **Version Updates**
  - Incremented version numbers for various modules to reflect recent changes.

### CODE REVIEW

**Comment by pipedream-component-development** [comment]
Thank you so much for submitting this! We've added it to our backlog to review, and our team has been notified.

**Comment by pipedream-component-development** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:
- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

**Review by luancazarine — APPROVED** [review]
Hi @philbuuza, Thanks for your contribution! LGTM! Ready for QA!

**Comment by philbuuza** [comment]
@luancazarine just did one little additional update to the overledger.app which will allow for instance selection between Sandox version and Live Overledger as they use different BaseUrls. once user selects an option in the props dropdown menu in environment prop this will set the correct baseURLs in the methods.

**Comment by philbuuza** [comment]
Please if you can approve after reviewing this so I can test it in the UI. Many thanks

File: components/overledger/overledger.app.mjs
Comment (philbuuza) [reply components/overledger/overledger.app.mjs:90]: apply these changes please

Comment (philbuuza) [reply components/overledger/overledger.app.mjs:90]: can you apply this change

File: components/overledger/actions/read-from-a-smart-contract/read-from-smart-contract.mjs
Comment (philbuuza) [reply components/overledger/actions/read-from-a-smart-contract/read-from-smart-contract.mjs:32]: I believe you can have object[] as a prop type according to the pipedream docs.  - https://pipedream.com/docs/components/api#user-input-props. 

Screenshot 2024-10-08 at 17 24 03

Comment (philbuuza) [reply components/overledger/actions/read-from-a-smart-contract/read-from-smart-contract.mjs:32]: hence why we have used string[] and parse these to create the required objects for the request

**Comment by philbuuza** [comment]
@luancazarine  I have made fixed all versioning and naming issues that were blocking the merge. Thanks

**Comment by philbuuza** [comment]
@luancazarine I have made the versioning alterations and would like for the merge to be done. many thanks

File: components/overledger/overledger.app.mjs
Comment (luancazarine) [inline components/overledger/overledger.app.mjs:43]: The method will receive the environment selected by the user
```suggestion
    _getBaseUrl(environment) { //conditional for environment url selection.
      return environment === "sandbox"
        ? "https://api.sandbox.overledger.io"
        : "https://api.overledger.io";
    },
```

File: components/overledger/overledger.app.mjs
Comment (luancazarine) [inline components/overledger/overledger.app.mjs:45]: Receiving the environment selected by the user
```suggestion
      $ = this, environment, path, ...otherOpts
```

File: components/overledger/overledger.app.mjs
Comment (luancazarine) [inline components/overledger/overledger.app.mjs:49]: Passing the environment selected by the user to _baseUrl method
```suggestion
        url: this._baseUrl(environment)+ path,
```

File: components/overledger/overledger.app.mjs
Comment (luancazarine) [inline components/overledger/overledger.app.mjs:56]: This repetition is unnecessary, the method already receives the environment in the opts param

File: components/overledger/overledger.app.mjs
Comment (luancazarine) [inline components/overledger/overledger.app.mjs:64]: This repetition is unnecessary, the method already receives the environment in the opts param

File: components/overledger/overledger.app.mjs
Comment (luancazarine) [inline components/overledger/overledger.app.mjs:72]: This repetition is unnecessary, the method already receives the environment in the opts param

File: components/overledger/overledger.app.mjs
Comment (luancazarine) [inline components/overledger/overledger.app.mjs:80]: This repetition is unnecessary, the method already receives the environment in the opts param

File: components/overledger/overledger.app.mjs
Comment (luancazarine) [inline components/overledger/overledger.app.mjs:90]: This repetition is unnecessary, the method already receives the environment in the opts param

File: components/overledger/overledger.app.mjs
Comment (luancazarine) [inline components/overledger/overledger.app.mjs:100]: This repetition is unnecessary, the method already receives the environment in the opts param

File: components/overledger/actions/read-from-a-smart-contract/read-from-a-smart-contract.mjs
Comment (luancazarine) [inline components/overledger/actions/read-from-a-smart-contract/read-from-a-smart-contract.mjs:71]:
```suggestion
      environment: this.environment,
      data: requestBody,
```

File: components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs
Comment (luancazarine) [inline components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs:66]:
```suggestion
      $,
      environment: this.environment,
```

**Review by luancazarine — CHANGES_REQUESTED** [review]
Hi @philbuuza, I just added some suggestions, please let me know if you have any questions.

You should add the environment prop in all actions and in the common source file and pass the environment prop in all run methods so that the application file can select the baseURL chosen by the user

File: components/overledger/overledger.app.mjs
Comment (philbuuza) [reply components/overledger/overledger.app.mjs:20]: this has been done - options has now been put in constants file as OVERLEDGER_INSTANCE

File: components/overledger/overledger.app.mjs
Comment (philbuuza) [reply components/overledger/overledger.app.mjs:6]: has been done. props has been removed and environment put into propDefinitions

File: components/overledger/actions/execute-signed-transaction/execute-signed-transaction.mjs
Comment (philbuuza) [reply components/overledger/actions/execute-signed-transaction/execute-signed-transaction.mjs:10]: has been done

File: components/overledger/actions/prepare-smart-contract-transaction/prepare-smart-contract-transaction.mjs
Comment (philbuuza) [reply components/overledger/actions/prepare-smart-contract-transaction/prepare-smart-contract-transaction.mjs:14]: done

File: components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs
Comment (philbuuza) [reply components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs:11]: also done now

File: components/overledger/actions/read-from-a-smart-contract/read-from-a-smart-contract.mjs
Comment (philbuuza) [reply components/overledger/actions/read-from-a-smart-contract/read-from-a-smart-contract.mjs:14]: also done now

File: components/overledger/overledger.app.mjs
Comment (philbuuza) [reply components/overledger/overledger.app.mjs:43]: change has been made

File: components/overledger/overledger.app.mjs
Comment (philbuuza) [reply components/overledger/overledger.app.mjs:49]: I assume here you meane to replace baseUrl with this._getBaseUrl(environment) and not this._baseUrl(environment).

File: components/overledger/overledger.app.mjs
Comment (philbuuza) [reply components/overledger/overledger.app.mjs:64]: agreed - has been removed.

File: components/overledger/overledger.app.mjs
Comment (philbuuza) [reply components/overledger/overledger.app.mjs:90]: within the "createHook" and "deletHook" methods do I just remove this param (baseUrl: this._getBaseUrl()) in the makeRequest methode call or do i replace here with sometthing else?

File: components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs
Comment (philbuuza) [reply components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs:66]: done

File: components/overledger/actions/read-from-a-smart-contract/read-from-a-smart-contract.mjs
Comment (philbuuza) [reply components/overledger/actions/read-from-a-smart-contract/read-from-a-smart-contract.mjs:71]: @luancazarine  done - But I assume this environment  should be added to the run function of both execute-signed-transaction and prepare-Smart-Contract-Transaction??

File: components/overledger/overledger.app.mjs
Comment (philbuuza) [reply components/overledger/overledger.app.mjs:100]: done

File: components/overledger/actions/read-from-a-smart-contract/read-from-a-smart-contract.mjs
Comment (philbuuza) [reply components/overledger/actions/read-from-a-smart-contract/read-from-a-smart-contract.mjs:71]: done

File: components/overledger/overledger.app.mjs
Comment (philbuuza) [reply components/overledger/overledger.app.mjs:80]: done

File: components/overledger/overledger.app.mjs
Comment (philbuuza) [reply components/overledger/overledger.app.mjs:49]: done

**Comment by philbuuza** [comment]
@luancazarine  - I have made the recommended changes as per your comments. Please review and merge if all ok. I did leave some reply comments on some of them but I seem to have clarified things now on further analysis

File: components/overledger/overledger.app.mjs
Comment (luancazarine) [reply components/overledger/overledger.app.mjs]: > I assume here you meant to replace baseUrl with this._getBaseUrl(environment) and not this._baseUrl(environment).

Yes, You're right

File: components/overledger/overledger.app.mjs
Comment (luancazarine) [reply components/overledger/overledger.app.mjs]: > within the "createHook" and "deletHook" methods do I just remove this param (baseUrl: this._getBaseUrl()) in the makeRequest methods call or do i replace here with sometthing else? bare in mind these methods are used in the sources/triggers (new-contract-event-instant and watch-new-account-instant) and not actions sections

@philbuuza, Yes, please remove the baseUrl, but make sure you have added the environment reference in the props of the common source file and passed it as a parameter in both methods

**Comment by luancazarine** [comment]
Hi @philbuuza. Thanks for making the changes I recommend. We're almost there.
I just need you to add the environment prop to the common source file and pass it to the create and delete hook methods.

**Comment by philbuuza** [comment]
I have added the environemnt to the common source file and pass it in the create and delete hook methods. Please review and see if this is ok. Many thanks for all your feedback

**Review by luancazarine — DISMISSED** [review]
Hi @philbuuza, thanks for your contribution! LGTM! Ready for QA!

**Comment by philbuuza** [comment]
Hi @luancazarine. Thanks for your help too. Is the final QA done by someone else in your team then. Is that the Vercel deploy thats waiting

**Comment by philbuuza** [comment]
@luancazarine  - I cant seem to find where the Type Error (this._sanboxBaseUrl is not a function) is coming from in the QA from @vunguyenhung. Its not a function we have in the code anywhere?. Am i missing something here?

**Comment by philbuuza** [comment]
Also can I confirm that apart from the issues wth read-from-smart-contract and sign-a-transantion all the other actions and sources have passed and ar ok?

**Comment by philbuuza** [comment]
have you got an Quant Connect account and created an app with client id and client key which are needed for Overledger to work

**Comment by philbuuza** [comment]
@vunguyenhung  Also looking at that function you are calling on that smart contract you need to provide input parameters for in. in this case it will be an address.:

//notice This is the test for making a call to a read function with a long function name
    function readVeryLongFunctionNameReadVeryLongFunctionNameReadVeryLongFunctionName(address account) public view returns (address) {
        return account;
    }

**Comment by philbuuza** [comment]
@luancazarine  @vunguyenhung  you can try a simplere one to begin with - whixh does not require inputs like  - readZeroInputsOneOutput. On the same smart contract address....

Also ca I check i there i anyway for me to do checkout to this PR branch and run the tests myself with the UI platform

**Comment by philbuuza** [comment]
Hi @vunguyenhung  ok sure. Also you can try a simpler function where you do not need to put any inputs and get a valued returned. like these:

1 - readZeroInputsOneOutput
2 - readZeroInputsTwoOutputs

Heres i the smart contract these are from. Where you can view the functions available:
https://sepolia.etherscan.io/address/0x3a637d769ed8530b483eeae70923d73dbad916c6#readContract

Please remember to have the sanbox instance of overledger set as well as selecting the Ethereum Sepolia Testnet.

**Comment by philbuuza** [comment]
have ammended the Sign-a-Transaction failed issue with the nativeData:

"Sign A Transaction - Fail
I got an error when publishing the action
bad default value for prop nativeData: undefined (should be object)"

Have added a default empty object to fix this undefined issue.

**Comment by philbuuza** [comment]
@luancazarine  and @vunguyenhung  I am still waiting on this PR. Can I get an update as to the progress of this. I dont have an upgraded account which would have allowed me to test this out myself so I kindly ask this is done so I can complete this task. Many thanks.

**Comment by philbuuza** [comment]
Hi @vunguyenhung  @luancazarine How do I test my action without being able to use the UI with it. Please if you could help me with this or direct me on how. would be much appreciated

File: components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs
Comment (luancazarine) [inline components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs:45]: It doesn't work
```suggestion
```

File: components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs
Comment (luancazarine) [inline components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs:23]: It doesn't work
```suggestion
```

**Review by luancazarine — CHANGES_REQUESTED** [review]
Hi @philbuuza, I just added two more suggestions.

File: components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs
Comment (philbuuza) [reply components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs:45]: Hi @luancazarine . Do you suggest I just remove this default value? Has it been tested with the Prep smart contract  step before hand?

File: components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs
Comment (philbuuza) [reply components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs:23]: Hi @luancazarine . Do you suggest I just remove this default value? Has it been tested with the Prep smart contract  step before hand?

File: components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs
Comment (luancazarine) [reply components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs:45]: Yes, just remove it. Props don't have access to steps, it will return error if you try to publish the component

Comment (philbuuza) [reply components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs]: done

File: components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs
Comment (philbuuza) [reply components/overledger/actions/sign-a-transaction/sign-a-transaction.mjs]: done

**Comment by luancazarine** [comment]
> @luancazarine any update on this. Can it be merged soon?

Hi @lukerQuant. Were you able to test the components as @vunguyenhung suggested?
Anyway, I'm moving the issue to the QA team

**Review by luancazarine — APPROVED** [review]

**Comment by lukerQuant** [comment]
@vunguyenhung

> > @luancazarine any update on this. Can it be merged soon?
> 
> Hi @lukerQuant. Were you able to test the components as @vunguyenhung suggested? Anyway, I'm moving the issue to the QA team

Hi @luancazarine , yes @philbuuza tested via the cli. We now look forward to @vunguyenhung's approval.

Best wishes

**Comment by philbuuza** [comment]
Hi @luancazarine  and @vunguyenhung  - Thanks for your time with this PR - as there has been alot of back and forth. All issues have now been sorted and fully tested. The only issue we can not seem to resolve but is not a real issue is function read function calls on functions that do not give an output as @vunguyenhung had discovered on his testing. Have tried all types all types of solutions and even omitting this parameter on the API request but nothing works. So we can just settle on this not being an issue as no one will really invoke read functions that do not return anything. This should not be an a blocker to pass the action. The outputParameter has been made optional and now has a default value of an empty  array []. The same has been done for input parameters which is now also and empty array as this always has to be sent in the API call. All other actions including sign transaction have passed. Please find attached some screenshots from my published actions via the CLI.
**(read from contract) -a**
a - number - read from smart contract - 1 of 4
a - number - read from smart contract - 2 of 4
a - number - read from smart contract - 3 of 4

**(read from contract) -b
b - readOneAInputOneAOutput - 1 of 4
b - readOneAInputOneAOutput - 3 of 4

read from contract - c
c - readOneUIInputsTwoUIOutputOneTwoEight - 1 of 4
c - readOneUIInputsTwoUIOutputOneTwoEight - 2 of 4
c - readOneUIInputsTwoUIOutputOneTwoEight - 3 of 4


prep Tranx
Prepare Tranx 1 of
Prepare Tranx 2 of
Prepare Tranx 3 of
Prepare Tranx 4 of

Sign tranx
Sign Tranx 1 of
Sign Tranx 2 of
Sign Tranx 3 of
Sign Tranx 4 of
Sign Tranx 5 of
Sign Tranx 6 of

Exec Tranx
exec Tranx 1 of
exec tranx 2 of
exec tranx 3 of

**Comment by philbuuza** [comment]
Hi @luancazarine. If we can get this merge done that would be great. @vunguyenhung has given the ok on the test.

**Review by luancazarine — APPROVED** [review]

**Comment by luancazarine** [comment]
Hi @philbuuza @lukerQuant! It's done! I've already merged it to master!!!

**Comment by philbuuza** [comment]
@luancazarine. Great stuff. many thanks for your input on this. Much appreciated

**Pull request merged**
